cmake/FindCoroutines.cmake

```diff
@@ -0,0 +1,287 @@
+  check_include_file_cxx("experimental/coroutine" _CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER)
+  if(NOT _CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER)
+    cmake_push_check_state()
+    set(CMAKE_REQUIRED_FLAGS "${_CXX_COROUTINES_EXTRA_FLAGS}")
+    check_include_file_cxx("experimental/coroutine" _CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER_WITH_FLAG)
+    set(_CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER "${_CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER_WITH_FLAG}")
+    cmake_pop_check_state()
+  endif()
+  mark_as_advanced(_CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER)
+else()
+  set(_CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER FALSE)
+endif()
+
+if(_CXX_COROUTINES_HAVE_HEADER)
+  set(_have_coro TRUE)
+  set(_coro_header coroutine)
+  set(_coro_namespace std)
+elseif(_CXX_COROUTINES_HAVE_EXPERIMENTAL_HEADER)
+  set(_have_coro TRUE)
+  set(_coro_header experimental/coroutine)
+  set(_coro_namespace std::experimental)
+else()
+  set(_have_coro FALSE)
+endif()
+
+set(CXX_COROUTINES_HAVE_COROUTINES ${_have_coro} CACHE BOOL "TRUE if we have the C++ coroutines feature")
+set(CXX_COROUTINES_HEADER ${_coro_header} CACHE STRING "The header that should be included to obtain the coroutines APIs")
+set(CXX_COROUTINES_NAMESPACE ${_coro_namespace} CACHE STRING "The C++ namespace that contains the coroutines APIs")
+
+set(_found FALSE)
+
+if(CXX_COROUTINES_HAVE_COROUTINES)
+  # We have some coroutines library available. Do link checks
+  string(CONFIGURE [[
+        #include <utility>
+        #include <@CXX_COROUTINES_HEADER@>
+
+        struct present {
+          struct promise_type {
+            int result;
+            present get_return_object() { return present{*this}; }
+            @CXX_COROUTINES_NAMESPACE@::suspend_never initial_suspend() { return {}; }
+            @CXX_COROUTINES_NAMESPACE@::suspend_always final_suspend() noexcept { return {}; }
+            void return_value(int i) { result = i; }
+            void unhandled_exception() {}
+          };
+          friend struct promise_type;
+          present(present&& that) : coro_(std::exchange(that.coro_, {})) {}
+          ~present() { if(coro_) coro_.destroy(); }
+          bool await_ready() const { return true; }
+          void await_suspend(@CXX_COROUTINES_NAMESPACE@::coroutine_handle<>) const {}
+          int await_resume() const { return coro_.promise().result; }
+        private:
+          present(promise_type& promise)
+            : coro_(@CXX_COROUTINES_NAMESPACE@::coroutine_handle<promise_type>::from_promise(promise)) {}
+          @CXX_COROUTINES_NAMESPACE@::coroutine_handle<promise_type> coro_;
+        };
+
+        present f(int n) {
+          if (n < 2)
+            co_return 1;
+          else
+            co_return n * co_await f(n - 1);
+        }
+
+        int main() {
+          return f(5).await_resume() != 120;
+        }
+    ]] code @ONLY)
+
+  # Try to compile a simple coroutines program without any compiler flags
+  set(CMAKE_EXE_LINKER_FLAGS_TEMP ${CMAKE_EXE_LINKER_FLAGS})
```

Comment:
<details><summary>Details</summary>
<p>
Save in temporary and clear 

</p>
</details>